### PR TITLE
Vda-web3-client Gas configuration updated

### PR DESCRIPTION
**Pull request by ITStar10**
Tested
	- for each gas configuration type

### CODE REVIEW

**Review by tahpot — COMMENTED** [review]
Two key changes:

1. Leave existing gas calculation code as a fallback if no gas config is provided
2. Gas limit per method is the units not the fee.

File: packages/vda-web3-client/src/VeridaContractBase.ts [hunk 1]
```diff
@@ -203,16 +209,29 @@ export class VeridaContract {
...
+                    // let gasLimit = await contract.estimateGas[methodName](...params);
+                    // gasLimit = gasLimit.mul(BigNumber.from(15)).div(BigNumber.from(10)) // Multiply 1.1
+                    // console.log(gasLimit.toString())
+
+                    const gasConfig : Record<string, any> = {}
+                    // {
+                    //     gasLimit: 500000 // gasLimit,
+                    //     // gasPrice
+                    // }
+                    if (this.web3Config?.fixedGasPerMethod?.has(methodName)) {
```
[the remark this answers is not in the document]
Comment (ITStar10) [reply packages/vda-web3-client/src/VeridaContractBase.ts:224]: They're used as a gasLimit already.
image

File: packages/vda-web3-client/src/VeridaContractBase.ts [hunk 2]
```diff
@@ -206,40 +209,45 @@ export class VeridaContract {
                 if (methodType === 'view') {
                     ret = await contract.callStatic[methodName](...params)
                 } else {
-                    if (this.logPerformance) {
-                        console.log('Node endpoint: ' + (<JsonRpcProvider> contract.provider).connection.url)
-                    }
+                    let gasConfig : Record<string, any> | undefined = {}
+                    // console.log("Params : ", params)
+
+                    const paramLen = params.length
```
Comment (tahpot) [inline packages/vda-web3-client/src/VeridaContractBase.ts:215]: Please specify a typescript type for `params`

File: packages/vda-web3-client/src/VeridaContractBase.ts [hunk 3]
```diff
@@ -206,40 +209,45 @@ export class VeridaContract {
...
+                    if (params !== undefined
```
Comment (tahpot) [inline packages/vda-web3-client/src/VeridaContractBase.ts:216]: This is very difficult to read.

Perhaps try doing something like:

const gasConfig = _.merge(params, this.methodDefaults[methodName], this.webConfig.gasConfig)

File: packages/vda-web3-client/test/didregistry.test.ts [hunk 4]
```diff
@@ -402,4 +402,47 @@ describe('DidRegistry Test', () => {
       );
     });
   });
+
+  describe('Manual gas configuration test', () => {
```
Comment (tahpot) [inline packages/vda-web3-client/test/didregistry.test.ts:406]: This is testing the gasConfig being specified as a param.

Can we have tests for `methodDefaults` and `web3config` settings?

**Review by tahpot — CHANGES_REQUESTED** [review]

File: packages/vda-web3-client/test/didregistry.test.ts (hunk 4 above)
Comment (ITStar10) [reply packages/vda-web3-client/test/didregistry.test.ts:406]: Global gas configuration & `methodDefaults` is already defiend in the "test/utils.ts" file at line#86 ~ line#95

Comment (tahpot) [reply packages/vda-web3-client/test/didregistry.test.ts]: They are defined there, but you need to tests that show the `web3config` is used if the other options aren't provided and that `methodDefaults` is used if gas config isn't specified in `params`.

**Review by tahpot — CHANGES_REQUESTED** [review]
Looks good, just one comment remaining on the test.

File: packages/vda-web3-client/test/didregistry.test.ts (hunk 4 above)
Comment (ITStar10) [reply packages/vda-web3-client/test/didregistry.test.ts]: Already tested for various cases. Please check `didregistry.test.ts` file again:
line#411 : `No gas configuration` is test for default web3 config. 
line#425 : `Call wigh global gas configuration` : Used the global gas cofiguration
line#459 : `Call with method default gas configuration` : Used the methodDefaults configuration with global configuration defined
line#513 : `Call with manual gas configuration` : Used to test gas configuration in the function parameters

Comment (tahpot) [reply packages/vda-web3-client/test/didregistry.test.ts]: Got it! I missed that. Looks great :)

**Review by tahpot — APPROVED** [review]

**Pull request merged**